src/theory/quantifiers/extended_rewrite.cpp

```diff
@@ -0,0 +1,331 @@
+
+#include "theory/datatypes/datatypes_rewriter.h"
+#include "theory/quantifiers/quant_util.h"  // for QuantArith
+#include "theory/rewriter.h"
+#include "theory/strings/theory_strings_rewriter.h"
+
+using namespace CVC4::kind;
+using namespace std;
+
+namespace CVC4 {
+namespace theory {
+namespace quantifiers {
+
+ExtendedRewriter::ExtendedRewriter()
+{
+  d_true = NodeManager::currentNM()->mkConst(true);
+  d_false = NodeManager::currentNM()->mkConst(false);
+}
+
+Node ExtendedRewriter::extendedRewritePullIte(Node n)
+{
+  // generalize this?
+  Assert(n.getNumChildren() == 2);
+  Assert(n.getType().isBoolean());
+  Assert(n.getMetaKind() != kind::metakind::PARAMETERIZED);
+  std::vector<Node> children;
+  for (unsigned i = 0; i < n.getNumChildren(); i++)
+  {
+    children.push_back(n[i]);
+  }
+  for (unsigned i = 0; i < 2; i++)
+  {
+    if (n[i].getKind() == kind::ITE)
+    {
+      for (unsigned j = 0; j < 2; j++)
+      {
+        children[i] = n[i][j + 1];
+        Node eqr = extendedRewrite(
+            NodeManager::currentNM()->mkNode(n.getKind(), children));
+        children[i] = n[i];
+        if (eqr.isConst())
+        {
+          std::vector<Node> new_children;
+          Kind new_k;
+          if (eqr == d_true)
+          {
+            new_k = kind::OR;
+            new_children.push_back(j == 0 ? n[i][0] : n[i][0].negate());
+          }
+          else
+          {
+            Assert(eqr == d_false);
+            new_k = kind::AND;
+            new_children.push_back(j == 0 ? n[i][0].negate() : n[i][0]);
+          }
+          children[i] = n[i][2 - j];
+          Node rem_eq = NodeManager::currentNM()->mkNode(n.getKind(), children);
+          children[i] = n[i];
+          new_children.push_back(rem_eq);
+          Node nc = NodeManager::currentNM()->mkNode(new_k, new_children);
+          Trace("sygus-ext-rewrite") << "sygus-extr : " << n << " rewrites to "
+                                     << nc << " by simple ITE pulling."
+                                     << std::endl;
+          // recurse
+          return extendedRewrite(nc);
+        }
+      }
+    }
+  }
+  return Node::null();
+}
+
+Node ExtendedRewriter::extendedRewrite(Node n)
+{
+  std::unordered_map<Node, Node, NodeHashFunction>::iterator it =
+      d_ext_rewrite_cache.find(n);
+  if (it == d_ext_rewrite_cache.end())
+  {
+    Node ret = n;
+    if (n.getNumChildren() > 0)
+    {
+      std::vector<Node> children;
+      if (n.getMetaKind() == kind::metakind::PARAMETERIZED)
+      {
+        children.push_back(n.getOperator());
+      }
+      bool childChanged = false;
+      for (unsigned i = 0; i < n.getNumChildren(); i++)
+      {
+        Node nc = extendedRewrite(n[i]);
+        childChanged = nc != n[i] || childChanged;
+        children.push_back(nc);
+      }
+      Node ret;
+      if (childChanged)
+      {
+        ret = NodeManager::currentNM()->mkNode(n.getKind(), children);
+      }
+    }
+    ret = Rewriter::rewrite(n);
```

Comment:
This should be applied on `ret`. Moreover, in the end of the `if` block above there is a shadowing of `ret`.

Reply:
Thanks, I'll fix.